File: Python3/프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
import itertools
import collections

def solution(friends, gifts):
    gift_list = collections.defaultdict(int)
    gift_point = collections.defaultdict(int)
    
    for gift in gifts:
        from_frd, to_frd = gift.split(" ")
        
        # 주고받은 선물 개수 계산
        gift_list[(from_frd, to_frd)] += 1
        
        # 선물지수 계산
        gift_point[from_frd] += 1
        gift_point[to_frd] -= 1
        
    result = collections.defaultdict(int)
    for frd_a, frd_b in itertools.combinations(friends, 2):
        if gift_list[(frd_a, frd_b)] < gift_list[(frd_b, frd_a)]:
            result[frd_b] += 1
        elif gift_list[(frd_a, frd_b)] > gift_list[(frd_b, frd_a)]:
            result[frd_a] += 1
        else:
            if gift_point[frd_a] < gift_point[frd_b]:
                result[frd_b] += 1
            elif gift_point[frd_a] > gift_point[frd_b]:
                result[frd_a] += 1

    return max(result.values()) if result else 0

File: Swift/프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.swift
import Foundation

func solution(_ friends:[String], _ gifts:[String]) -> Int {
    var giftList = [GiftCase: Int]()
    var giftPoint = [String: Int]()
    
    for gift in gifts {
        let frds = gift.split(separator: " ")
        let from_frd = String(frds[0])
        let to_frd = String(frds[1])
        
        // 선물 주고받은 내역 계산
        giftList[GiftCase(from_frd, to_frd), default: 0] += 1
        
        // 선물지수 계산
        giftPoint[from_frd, default: 0] += 1
        giftPoint[to_frd, default: 0] -= 1
    }
    
    var cnt = friends.count
    var result = [String: Int]()
    
    for (a, frd_a) in friends.enumerated() {
        for b in a+1..<cnt {
            let frd_b = friends[b]
            
            if giftList[GiftCase(frd_a, frd_b)] ?? 0 < giftList[GiftCase(frd_b, frd_a)] ?? 0 {
                result[frd_b, default: 0] += 1
            } else if giftList[GiftCase(frd_a, frd_b)] ?? 0 > giftList[GiftCase(frd_b, frd_a)] ?? 0 {
                 result[frd_a, default: 0] += 1
            } else {
                // 주고받은게 같으면 선물지수 비교
                if giftPoint[frd_a] ?? 0 < giftPoint[frd_b] ?? 0 {
                    result[frd_b, default: 0] += 1
                } else if giftPoint[frd_a] ?? 0 > giftPoint[frd_b] ?? 0 {
                    result[frd_a, default: 0] += 1
                }
            }
        }
    }
    
    if let ans = result.values.max() {
        return ans
    } else {
        return 0
    }
}



struct GiftCase: Hashable {
    var to_frd: String
    var from_frd: String
    
    init(_ from: String, _ to: String) {
        to_frd = to
        from_frd = from
    }
}